lib/internal/test_runner/test.js

```diff
@@ -0,0 +1,396 @@
+    this.startTime = hrtime();
+
+    try {
+      await this.runInAsyncScope(this.fn, null, new TestContext(this));
+      this.pass();
+    } catch (err) {
+      this.fail(new ERR_TEST_FAILURE(err, kTestCodeFailure));
+    }
+
+    // Clean up the test. Then, try to report the results and execute any
+    // tests that were pending due to available concurrency.
+    this.postRun();
+  }
+
+  postRun() {
+    let failedSubtests = 0;
+
+    // If the test was failed before it even started, then the end time will
+    // be earlier than the start time. Correct that here.
+    if (this.endTime < this.startTime) {
+      this.endTime = hrtime();
+    }
+
+    // The test has run, so recursively cancel any outstanding subtests and
+    // mark this test as failed if any subtests failed.
+    for (let i = 0; i < this.subtests.length; i++) {
+      const subtest = this.subtests[i];
+
+      if (subtest.endTime === null) {
+        subtest.cancel();
+        subtest.postRun();
+      }
+
+      if (!subtest.passed) {
+        failedSubtests++;
+      }
+    }
+
+    if (this.passed && failedSubtests > 0) {
+      const subtestString = `subtest${failedSubtests > 1 ? 's' : ''}`;
+      const msg = `${failedSubtests} ${subtestString} failed`;
+
+      this.fail(new ERR_TEST_FAILURE(msg, kSubtestsFailed));
+    }
+
+    if (this.parent !== null) {
+      this.parent.activeSubtests--;
+      this.parent.addReadySubtest(this);
+      this.parent.processReadySubtestRange(false);
+      this.parent.processPendingSubtests();
+    }
+  }
+
+  isClearToSend() {
+    return this.parent === null ||
+      (
+        this.parent.waitingOn === this.testNumber && this.parent.isClearToSend()
+      );
+  }
+
+  finalize() {
+    // By the time this function is called, the following can be relied on:
+    // - The current test has completed or been cancelled.
+    // - All of this test's subtests have completed or been cancelled.
+    // - It is the current test's turn to report its results.
+
+    // Report any subtests that have not been reported yet. Since all of the
+    // subtests have finished, it's safe to pass true to
+    // processReadySubtestRange(), which will finalize all remaining subtests.
+    this.processReadySubtestRange(true);
+
+    // Output this test's results and update the parent's waiting counter.
+    if (this.subtests.length > 0) {
+      this.reporter.plan(this.subtests[0].indent, this.subtests.length);
+    }
+
+    this.report();
+    this.parent.waitingOn++;
+
+    if (this.parent.parent === null &&
+        this.parent.waitingOn > this.parent.subtests.length) {
+      // This was the last test, so the root does not need to keep the event
+      // loop alive anymore.
+      clearInterval(this.parent.keepAlive);
+    }
+  }
+
+  report() {
+    // Duration is recorded in BigInt nanoseconds. Convert to seconds.
+    const duration = Number(this.endTime - this.startTime) / 1_000_000_000;
+    const message = `- ${this.name}`;
```

Comment:
Is the `- ` necessary? Other test harnesses will not prefix the names and `TAP` output seems cleaner without it?!

Reply:
It's not strictly necessary. I've seen other test runners (including the one core currently uses and node-tap) add it.

Reply:
It looks like the specification for TAP 14 does explicitly note the `-` being treated as optional and OK: https://github.com/TestAnything/Specification/pull/25/files#diff-1fdb6f17d18295d39b8a3ee5b260374867993b9306f3130b7df76627c60615c7R321-R324